include/drivers/dai.h

```diff
@@ -0,0 +1,629 @@
+ *  NB represents "Normal Bit Clk" whereas IB represents "Inverted Bit clk"
+ */
+#define DAI_FMT_CLK_NF_NB		(0 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_NF_IB		(1 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_IF_NB		(2 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_IF_IB		(3 << DAI_FMT_CLK_FORMAT_SHIFT)
+
+typedef uint8_t dai_opt_t;
+
+/** Run bit clock continuously */
+#define DAI_OPT_BIT_CLK_CONT                (0 << 0)
+/** Run bit clock when sending data only */
+#define DAI_OPT_BIT_CLK_GATED               BIT(0)
+
+/** DAI driver is bit clock provider (codec is consumer) */
+#define DAI_OPT_BIT_CLK_PROVIDER              (0 << 1)
+/** DAI driver is bit clock consumer (codec is provider) */
+#define DAI_OPT_BIT_CLK_CONSUMER               BIT(1)
+/** DAI driver is frame clock provider (codec is consumer) */
+#define DAI_OPT_FRAME_CLK_PROVIDER            (0 << 2)
+/** DAI driver is frame clock consumer (codec is provider) */
+#define DAI_OPT_FRAME_CLK_CONSUMER             BIT(2)
+
+/** @brief Loop back mode.
+ *
+ * In loop back mode RX input will be connected internally to TX output.
+ * This is used primarily for testing.
+ */
+#define DAI_OPT_LOOPBACK                    BIT(7)
+
+/** @brief Ping pong mode
+ *
+ * Inherited from legacy TBD to check.
+ *
+ * In ping pong mode TX output will keep alternating between a ping buffer
+ * and a pong buffer. This is normally used in audio streams when one buffer
+ * is being populated while the other is being played (DMAed) and vice versa.
+ * So, in this mode, 2 sets of buffers fixed in size are used. Static Arrays
+ * are used to achieve this and hence they are never freed.
+ */
+#define DAI_OPT_PINGPONG                    BIT(6)
+
+/** \brief Types of DAI
+ *
+ * The type of the DAI. This ID type is used to configure bespoke DAI HW
+ * settings.
+ *
+ * DAIs have a lot of physical link feature variability and therefore need
+ * different configuration data to cater for different use cases. We
+ * usually need to pass extra bespoke configuration prior to DAI start.
+ */
+enum dai_type {
+	DAI_LEGACY_I2S = 0,	/**< Legacy I2S compatible with i2s.h */
+	DAI_INTEL_SSP,		/**< Intel SSP */
+	DAI_INTEL_DMIC,		/**< Intel DMIC */
+	DAI_INTEL_HDA,		/**< Intel HD/A */
+	DAI_INTEL_ALH,		/**< Intel ALH */
+	DAI_IMX_SAI,                /**< i.MX SAI */
+	DAI_IMX_ESAI,               /**< i.MX ESAI */
+	DAI_AMD_BT,			/**< Amd BT */
+	DAI_AMD_SP,			/**< Amd SP */
+	DAI_AMD_DMIC,		/**< Amd DMIC */
+	DAI_MEDIATEK_AFE,            /**< Mtk AFE */
+	DAI_INTEL_SSP_NHLT,            /**< nhlt ssp */
+	DAI_INTEL_DMIC_NHLT,            /**< nhlt ssp */
+	DAI_INTEL_HDA_NHLT,		/**< nhlt Intel HD/A */
+	DAI_INTEL_ALH_NHLT,		/**< nhlt Intel ALH */
+};
+
+/**
+ * @brief Dai Direction
+ * TODO codec vs soc direction
+ */
+enum dai_dir {
+	/** Receive data */
+	DAI_DIR_RX = 0,
+	/** Transmit data */
+	DAI_DIR_TX,
+	/** Both receive and transmit data */
+	DAI_DIR_BOTH,
```

Comment:
Easier to validate a single direction capability when using flags (rx=1, tx=2, both=(rx|tx))

Reply:
I think this replaced old interface, so we don't need to stay compatible. It's similar enough, so porting is still quite easy.

Reply:
ok